core/store/src/main/java/org/locationtech/geowave/core/store/query/aggregate/BinningAggregation.java

```diff
@@ -0,0 +1,134 @@
+package org.locationtech.geowave.core.store.query.aggregate;
+
+import com.google.common.collect.Maps;
+import org.locationtech.geowave.core.index.StringUtils;
+import org.locationtech.geowave.core.index.persist.Persistable;
+import org.locationtech.geowave.core.store.api.Aggregation;
+import java.nio.ByteBuffer;
+import java.util.Map;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+public class BinningAggregation<P extends Persistable, R, T> implements
+    Aggregation<P, Map<String, R>, T> {
+
+  private Supplier<Aggregation<P, R, T>> baseAggregation;
+  private AggregationBinningStrategy<T> binningStrategy;
+  private Map<String, Aggregation<P, R, T>> result;
+
+  // TODO: Add this Agg to the adapters list? Its only used internally, so unneeded?
+  public BinningAggregation() {
+    this(null, null);
+  }
+
+  public BinningAggregation(
+      Supplier<Aggregation<P, R, T>> baseAggregation,
+      AggregationBinningStrategy<T> binningStrategy) {
+    this.baseAggregation = baseAggregation;
+    this.binningStrategy = binningStrategy;
+    this.result = Maps.newHashMapWithExpectedSize(1024);
```

Comment:
I changed this already. The BinningStrategy now has a 'sizeHint' default method that an impl can override to give a better estimate. The default returns `100` which is a decidedly non-scientific number.

Reply:
I reverted the addition of the `sizeHint` default method., because it can return pretty huge values. Precision 6 geohashes can have around 1Billion values, so the size hint would return that. It caused OOMs in my testing (of course making a 1-billion size hashmap would do that :P)

Any recommendations on what to do? Just using the default `10` of a hashmap probably isnt a good idea.

Reply:
Well wouldn't the user be using the `maxBinsAllowed` in the constructor for the GeoHashBinningStrategy?  They would have to have enough memory for 1 billion bins if they specified that many allowed.

Reply:
So, I think that there are valid reasons to be using a precision of 6 (~1B hashes), while only needing a small fraction of the total geohashes that that precision gives.

For example, If I had data of only 1 small country, but it was fairly fine grained, then using precision 6 may be warranted. I may only need ~1 million geohashes of precision 6 to cover the entire country. If I used sizeHint as it I implemented it, it would preallocate a map of 1B elements, which causes an easy OOM.

Under the current system, you may say 'just say maxBins 1,000,000'. But when I do that, I would only get a precision of `4`, which may not be fine-grained enough for the data.

This doesnt solve the preallocation problem, but I think that using maxBins is actually a poor choice for Geohashes, because its global-scale. Also people using geohashes are used to precision, and not used to an arbitrary 'max bins' calculating. If we attached some sort of BBox and did some calculations to figure out, given the BBox and the max bins, what precision can we use, but thats just getting really complicated for geohashes, which are pretty straightforward. 

Im not a geospatial guy, really, so I could be completely off the mark. My experience with geohashes has been that people understand the precision system fairly well. Sorry this became a mini rant cause I dont think I like the maxBins system that much. Of course I dont mind being told 'f-off we are gonna use maxBins' because again, I dont really know what im talking about.